### PR TITLE
LSP: Register keymaps on `LspAttach`

### LINKED ISSUE #1219: feature: register LSP binds on `LspAttach`
Currently, LSP's keymap registration is done together with most other keymaps, as LSP keymaps actually get added to `programs.nixvim.keymaps` after a bit of processing.

It's common practice in neovim configs to only register LSP-related keymaps when the buffer is attached to a language server, so that vanilla keys can be overridden with LSP functions (etc).

This is demonstrated by [kickstart.nvim here](https://github.com/nvim-lua/kickstart.nvim/blob/66e2a5a425a4f3b2e4f9456d82b29718c24a8993/init.lua#L440-L446).

- This could be configured using a new option such as `plugins.lsp.keymaps.onAttach`

If this is supported, users may wish to add arbitrary keymaps when the LSP is attached, not just the `vim.lsp.*` functions. E.g. to make use of [telescope's lsp functions](https://github.com/nvim-telescope/telescope.nvim?tab=readme-ov-file#neovim-lsp-pickers).

- This could be configured by adding an "extra" list; `plugins.lsp.keymaps.extra`, which is a list of normal keymap submodules.
- Alternatively, users could add keymaps themselves, and manually set the keymap event to `"LspAttach"`.
- Ideally, both configuration methods would be available.

### Potential strategy

One way to approach this might be to add a "registered on event" attribute to the keymaps submodule...

When rendering keymaps to lua, first _grouped them_ using the new `event` attribute.

The "`null`" group would be registered the same way keymaps are now, since it represents keymaps with no `event`.

All other groups would be mapped to an `autoCmd` submodule, with `event` being the group's key and the `callback` function registering the group's keymaps like normal.

The LSP plugin could then take advantage of the new system by having it's keymaps use `event = "LspAttach"` by default. Similar to how `options.silent` is configured currently.

Draft implementation of grouping keymaps by event in [this comment](https://github.com/nix-community/nixvim/issues/1219#issuecomment-1984855570) below.

### PR DESCRIPTION
**Pull request by Bodleum**
Register keymaps defined in `plugins.lsp.keymaps` on `LspAttach`.
Added `plugins.lsp.keymaps.extra` to register arbitrary keymaps.

Fixes #1219

### CODE REVIEW

File: plugins/lsp/default.nix
Comment (MattSturgeon) [inline plugins/lsp/default.nix:161]: I wonder if this would add more confusion, since most users will use the "normal" hover?

File: plugins/lsp/default.nix
Comment (MattSturgeon) [inline plugins/lsp/default.nix:280]: If this line is being indented anyway, is it worth changing it to use `helpers.mkRaw`?

File: plugins/lsp/default.nix
Comment (MattSturgeon) [inline plugins/lsp/default.nix:264]: Could avoid indenting the lines below by doing something like this:
```suggestion
      keymapsOnEvents."LspAttach" = let
```

**Review by MattSturgeon — APPROVED** [review]
Looks good. Left a few comments to probe discussion.

File: plugins/lsp/default.nix
Comment (Bodleum) [reply plugins/lsp/default.nix:161]: Yes, I just added that, and the telescope mapping as examples of why you would use this as well as how.

Easily removed if you think that'd be clearer.

Comment (MattSturgeon) [reply plugins/lsp/default.nix:161]: No strong opinions either way. More examples is usually good, this one just felt like it could potentially confuse new users.

I'll let the maintainers decide if it should be changed.

**Comment by MattSturgeon** [comment]
Probably a good idea to update the description for `diagnostics` and `lspBuf` mentioning that the keymaps will be registered on `LspAttach`.
Similar to your description for `extra`.

File: plugins/lsp/default.nix
Comment (MattSturgeon) [reply plugins/lsp/default.nix:164]: I think this is fine now, with the updated description.

**Review by MattSturgeon — APPROVED** [review]
LGTM

**Review by GaetanLepage — APPROVED** [review]
LGTM !
Thanks @Bodleum for this work and @MattSturgeon for kindly reviewing it.

As this PR affects the LSP code, I will wait for an extra review from @traxys before merging.

**Review by traxys — APPROVED** [review]

**Pull request merged**
